content/software-testing-the-hardware-software-interface-part-2.md

```diff
@@ -0,0 +1,417 @@
+import time
+import xmlrunner
+
+
+class SerialPromptTest(unittest.TestCase):
+    SERIAL_PORT = "/dev/tty.usbserial-110085063"
+    BAUD_RATE = 115200
+    TIMEOUT = 40
+    PROMPT = "verdin-imx8mp-06817296 login:"
+
+    def test_wait_for_PROMPT(self):
+        try:
+            with serial.Serial(
+                self.SERIAL_PORT, self.BAUD_RATE, timeout=self.TIMEOUT
+            ) as ser:
+                time.sleep(1)
+                ser.reset_input_buffer()
+
+                start_time = time.time()
+                received_data = ""
+
+                while time.time() - start_time < self.TIMEOUT:
+                    if ser.in_waiting > 0:
+                        chunk = ser.read(ser.in_waiting).decode(errors="ignore")
+                        print(chunk, end="")
+                        received_data += chunk
+
+                        if self.PROMPT in received_data:
+                            print(f"\nFound expected prompt '{self.PROMPT}'")
+                            return
+                    else:
+                        time.sleep(0.1)
+
+                self.fail(
+                    f"Expected prompt '{self.PROMPT}' not found within {self.TIMEOUT} seconds."
+                )
+
+        except serial.SerialException as e:
+            self.fail(f"Serial connection failed: {e}")
+
+
+if __name__ == "__main__":
+    with open("serial_test_results.xml", "wb") as output:
+        unittest.main(
+            testRunner=xmlrunner.XMLTestRunner(output=output), exit=False
+        )
+```
+
+The `test_wait_for_PROMPT` is quite simple and can be reutilized many times. The basic gist is that we keep looking at
+data chunks coming from the UART, decoding them and waiting until our desired prompt (`verdin-imx8mp-06817296 login:`)
+shows up.
+
+Because this function is encapsulated into a module that calls `unittest.main`, it can output the tests results in a
+standardized format for us, JUnit.xml, which can be interpreted by many other services such as GitLab, GitHub and others.
+
+Another options is using the handy `junit2html` program, which renders the test results as easily digestible HTML pages
+that can be viewed with a web browser.
+
+{{< figure src="/software-testing-part-2-junit2html.png" title="junit2html result of our automated test" alt="" width="50%" class="centered-figure">}}
+
+
+It might seem like... not a lot, right? But one can easily integrate this further by using `ser2net`, which takes a
+serial output and exposes it as a telnet connection for a distributed testing environment.
+
+To use `ser2net` I simply installed on my Debian machine using `apt install ser2net` and I'm running manually with
+
+`# ser2net -n -c /etc/ser2net.conf -P /run/ser2net.pid`
+
+where `/etc/ser2net.conf` contains the following
+
+`9000:raw:0:/dev/ttyUSB0:115200 8DATABITS NONE 1STOPBIT`.
+
+With that, from anywhere in the world, I can connect to my Verdin's UART by using telnet. The diagram looks a bit like
+this now:
+
+```
```

Comment:
not setting the type here is rendering the text as highlighted, which is odd IMHO.

Reply:
in fact, all the code blocks without the type are rendering in the same way, please don't be lazy and render the page before you send the PR.

Reply:
Fixed, thanks.